### PR TITLE
Add support for PrivateDnsZone

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

Adds support for the resource `PrivateDnsZone`.

This resource is a bit odd, as it has no properties of it's own, but exists more as something for other resources to reference. It seems to only exist in the `2018-09-01` version, so the imports needed to be carefully scoped to avoid importing other resources that we can't handle.

**Special notes for your reviewer**:

Will allow us to close PR #2086 by providing the requested resource, ahead of Swagger-only support.

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

File: v2/internal/testcommon/kube_per_test_context.go [hunk 1]
```diff
@@ -602,21 +602,46 @@ func (tc *KubePerTestContext) AsExtensionOwner(obj client.Object) *genruntime.Ar
...
 
 	tc.T.Logf("Exported resource to %s", filepath)
 }
 
+func (tc *KubePerTestContext) cleanSample(resource any) {
+
+	if kr, ok := resource.(genruntime.KubernetesResource); ok {
+		// Remove Status
+		emptyStatus := kr.NewEmptyStatus()
+		_ = kr.SetStatus(emptyStatus) // Ignore errors
```
Comment (super-harsh) [inline v2/internal/testcommon/kube_per_test_context.go:627]: do we need empty status here or nil status here? As seen in the sample `status: {}` is present.

File: v2/config/samples/network/v1beta/v1beta20180901_privatednszone.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+apiVersion: network.azure.com/v1beta20180601
+kind: PrivateDnsZone
+metadata:
+  creationTimestamp: null
```
Comment (super-harsh) [inline v2/config/samples/network/v1beta/v1beta20180901_privatednszone.yaml:4]: Do we require this and empty status below?

**Review by super-harsh — CHANGES_REQUESTED** [review]
Resource addition LGTM, left a few comments for sample. Also, you might need to run `TEST_FILTER=Test_Samples_CreationAndDeletion task controller:test-integration-envtest` to make sure sample is tested and recording is added for the new sample.

On the above, its my bad I did not add step for recording samples test in documentation. Have created an issue #2433 to track it.

File: v2/internal/testcommon/kube_per_test_context.go (hunk 1 above)
Comment (theunrepentantgeek) [reply v2/internal/testcommon/kube_per_test_context.go:627]: We actually want no status at all; will modify.

File: v2/config/samples/network/v1beta/v1beta20180901_privatednszone.yaml (hunk 2 above)
Comment (theunrepentantgeek) [reply v2/config/samples/network/v1beta/v1beta20180901_privatednszone.yaml:4]: Not required; will remove.

File: v2/config/samples/network/v1beta/v1beta20180901_privatednszone.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+apiVersion: network.azure.com/v1beta20180601
```
Comment (super-harsh) [inline v2/config/samples/network/v1beta/v1beta20180901_privatednszone.yaml:1]:
```suggestion
apiVersion: network.azure.com/v1beta20180901
```

I just noticed failed on your PR.. ​runtime.notRegisteredErr​error indicates about either either of gvk in the sample is not correct. This should fix the error.

Comment (theunrepentantgeek) [reply v2/config/samples/network/v1beta/v1beta20180901_privatednszone.yaml]: Thanks. I've just merged the same fix into my local branch.

File: v2/internal/controllers/crd_networking_privatednszone_test.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+			Owner:    testcommon.AsOwner(rg),
+		},
+	}
+
+	tc.CreateResourceAndWait(zone)
+
+	tc.Expect(zone.Status.MaxNumberOfRecordSets).ToNot(BeNil())
+	tc.Expect(zone.Status.MaxNumberOfVirtualNetworkLinks).ToNot(BeNil())
+
+	tc.DeleteResourceAndWait(zone)
```
Comment (matthchr) [inline v2/internal/controllers/crd_networking_privatednszone_test.go:38]: Minor: Could add assertion after this that delete really happened in Azure. Look for example tests that use the `HEAD` API to do that. For example the `crd_cache_redis_test.go` does that

File: v2/internal/testcommon/kube_per_test_context.go [hunk 5]
```diff
@@ -602,21 +602,46 @@ func (tc *KubePerTestContext) AsExtensionOwner(obj client.Object) *genruntime.Ar
...
+	}
+
+	if oa, ok := resource.(metav1.ObjectMetaAccessor); ok {
+		// Remove runtime objectmeta information
+		om := oa.GetObjectMeta()
+		om.SetAnnotations(nil)
+		om.SetFinalizers(nil)
+		om.SetManagedFields(nil)
+		om.SetLabels(nil)
+		om.SetOwnerReferences(nil)
```
Comment (matthchr) [inline v2/internal/testcommon/kube_per_test_context.go:637]: Consider setting namespace to `default` as well.

File: v2/internal/controllers/crd_networking_privatednszone_test.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+	tc := globalTestContext.ForTest(t)
+
+	rg := tc.CreateTestResourceGroupAndWait()
+
+	name := tc.Namer.GenerateName("pdz") + ".com"
+	zone := &network.PrivateDnsZone{
+		ObjectMeta: tc.MakeObjectMetaWithName(name),
+		Spec: network.PrivateDnsZones_Spec{
+			Location: to.StringPtr("global"),
```
Comment (matthchr) [inline v2/internal/controllers/crd_networking_privatednszone_test.go:28]: This is an interesting "location" that I haven't seen before. Do you know what it means?

**Review by matthchr — APPROVED** [review]

File: v2/internal/controllers/crd_networking_privatednszone_test.go (hunk 4 above)
Comment (theunrepentantgeek) [reply v2/internal/controllers/crd_networking_privatednszone_test.go:38]: Checking against Azure requires an ID or something to query for, and there's nothing on the object. Other tests use an Id property that doesn't exist here.

File: v2/internal/controllers/crd_networking_privatednszone_test.go (hunk 6 above)
Comment (theunrepentantgeek) [reply v2/internal/controllers/crd_networking_privatednszone_test.go:28]: DNS settings aren't (cannot be!) scoped to a region, so while all ARM resources need a location, `PrivateDNSZone` doesn't use it - and seems to require `global` as the value.

File: v2/internal/controllers/crd_networking_privatednszone_test.go (hunk 4 above)
Comment (matthchr) [reply v2/internal/controllers/crd_networking_privatednszone_test.go:38]: We also compute the Azure ID and store it in an annotation: `serviceoperator.azure.com/resource-id`, which is accessible as `genruntime.ResourceIDAnnotation` or readable via `genruntime.GetResourceID(obj)`

Comment (theunrepentantgeek) [reply v2/internal/controllers/crd_networking_privatednszone_test.go:38]: Added.

File: v2/config/samples/network/v1beta/v1beta20180901_privatednszone.yaml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+apiVersion: network.azure.com/v1beta20180901
+kind: PrivateDnsZone
+metadata:
+  name: aso-sample-pdz.com
+  namespace: aso-test-networking-privatednszone-crud
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply v2/config/samples/network/v1beta/v1beta20180901_privatednszone.yaml:5]: Done.

File: v2/internal/testcommon/kube_per_test_context.go (hunk 5 above)
Comment (theunrepentantgeek) [reply v2/internal/testcommon/kube_per_test_context.go:637]: Done.

**Review by super-harsh — APPROVED** [review]

**Pull request merged**